android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java

```diff
@@ -64,6 +76,131 @@ public void beepKeyboard() {
+   */
+  @JavascriptInterface
+  public void insertText(final int dn, final String s, final int dr, final boolean executingHardwareKeystroke) {
+    Handler mainLoop = new Handler(Looper.getMainLooper());
+    mainLoop.post(new Runnable() {
+      public void run() {
+        if (k == null) {
+          KMLog.LogError(TAG, "insertText failed: Keyboard is null");
+          return;
+        }
+
+        if (k.subKeysWindow != null) {
+          return;
+        }
+
+        if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP &&
+          (KMTextView.activeView == null || KMTextView.activeView.getClass() != KMTextView.class)) {
+          if (KMTextView.activeView == null && !KMManager.isDebugMode()) {
+            Log.w(TAG, "insertText failed: activeView is null");
+          }
+          return;
+        }
+
+        InputConnection ic = (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP) ?
+            KMTextView.activeView.onCreateInputConnection(new EditorInfo()) :
+            KMManager.getInputMethodService().getCurrentInputConnection();
+        if (ic == null) {
+          if (KMManager.isDebugMode()) {
+            Log.w(TAG, "insertText failed: InputConnection is null");
+          }
+          return;
+        }
+
+        ic.beginBatchEdit();
+
+        int deleteLeft = dn;
+
+        // Delete any existing selected text.
+        ExtractedText icText = ic.getExtractedText(new ExtractedTextRequest(), 0);
+        if (icText != null) { // This can be null if the input connection becomes invalid.
+          int start = icText.startOffset + icText.selectionStart;
+          int end = icText.startOffset + icText.selectionEnd;
+          if (end < start) {
+            // Swap start/end for backward selection
+            int temp = start;
+            start = end;
+            end = temp;
+          }
+          if (end > start) {
+            if (s.length() == 0) {
+              ic.setSelection(start, start);
+              ic.deleteSurroundingText(0, end - start);
+              ic.endBatchEdit();
+              return;
+            } else {
+              if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_SYSTEM) {
+                KMManager.SystemKeyboardShouldIgnoreSelectionChange = true;
+              }
+              ic.setSelection(start, start);
+              ic.deleteSurroundingText(0, end - start);
+            }
+
+            // KeymanWeb tells us how to delete the selection, but we don't
+            // want to do that twice
+            deleteLeft = 0;
+          }
+        }
+
+        if (s.length() > 0 && s.charAt(0) == '\n') {
+          if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP) {
+            KMTextView textView = (KMTextView)KMTextView.activeView;
+            textView.keyDownUp(KeyEvent.KEYCODE_ENTER);
+          } else if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_SYSTEM) {
+            keyDownUp(KeyEvent.KEYCODE_ENTER);
+          }
```

Comment:
Could this be put into `keyDownUp` itself? Otherwise `keyDownUp` is only appropriate for `KEYBOARD_TYPE_SYSTEM` from what I can tell, and would need to be named appropriately.

Reply:
ok, I've moved it into `keyDownUp`